lib/src/main/kotlin/com/swmansion/starknet/provider/Provider.kt

```diff
@@ -202,66 +204,112 @@ interface Provider {
     }
 
     /**
-     * Deploy a contract
+     * Declare contract
      *
-     * Deploy a contract on StarkNet.
+     * Declare a contract on StarkNet.
      *
-     * @param payload deploy transaction payload
+     * @param payload declare transaction payload
      *
      * @throws RequestFailedException
      */
-    fun deployContract(payload: DeployTransactionPayload): Request<DeployResponse>
+    fun declareContract(payload: DeclareTransactionPayload): Request<DeclareResponse>
 
     /**
-     * Declare contract
+     * Estimate a fee.
      *
-     * Declare a contract on StarkNet.
+     * Estimate a fee for a provided transaction.
      *
-     * @param payload declare transaction payload
+     * @param payload deploy account transaction, for which the fee is to be estimated.
+     * @param blockHash a hash of the block in respect to what the query will be made
      *
      * @throws RequestFailedException
      */
-    fun declareContract(payload: DeclareTransactionPayload): Request<DeclareResponse>
+    fun getEstimateFee(payload: DeployAccountTransactionPayload, blockHash: Felt): Request<EstimateFeeResponse>
```

Comment:
Since the payloads differ, I figured it should be done this way (like in the example of BlockId, that is being split in interface to indicate Felt/Hash/Tag). Do you think it is a good practice to make "an abstraction" of some kind over a type in the interface?

Reply:
I'm not sure why BlockId was split, maybe it didn't behave nicely in Java? I think that having function accept any kind of payload (perhaps you'd have to use either `out` or `in` keywords too but im not sure) i.e. the payload base class would be pretty standard.

Reply:
But perhaps that's for another PR before this get's released? I'm not sure.

Reply:
@Solpatium would love to hear what you think about it, cuz I'm not sure how to handle it as well

Reply:
This doesn't mean that we can't accept all kinds of payloads in a single method, the problem with blockId was different - we want to provide api using primitives, not wrapped types and here we are using our classes which have a nice hierarchy already. I think accepting any payload would simplify the flow.

Reply:
https://github.com/software-mansion/starknet-jvm/issues/224